gobblin-data-management/src/main/java/org/apache/gobblin/data/management/dataset/TimePartitionGlobFinder.java

```diff
@@ -0,0 +1,148 @@
+
+  private static final String DEFAULT_TIME_ZONE = "America/Los_Angeles";
+
+  private final String datasetPattern;
+  private final String datasetPartitionPattern;
+  private final Path yesterdayPartition;
+
+  private final Properties props;
+  private final FileSystem fs;
+  private final boolean enableEmptyPartition;
+
+  public TimePartitionGlobFinder(FileSystem fs, Properties properties) {
+    datasetPattern = properties.getProperty(ConfigurableGlobDatasetFinder.DATASET_FINDER_PATTERN_KEY);
+    Path datasetPartitionPath = new Path(datasetPattern);
+
+    String partitionPrefix = properties.getProperty(PARTITION_PREFIX, "");
+    String timeFormat = properties.getProperty(TIME_FORMAT).trim();
+    DateTimeFormatter timeFormatter = DateTimeFormatter.ofPattern(timeFormat);
+    datasetPartitionPattern = new Path(datasetPartitionPath,
+        partitionPrefix + derivePartitionPattern(timeFormat)).toString();
+    log.info("Dataset partition pattern is {}", datasetPartitionPattern);
+
+    ZonedDateTime curTime = ZonedDateTime.now(ZoneId.of(properties.getProperty(TIME_ZONE, DEFAULT_TIME_ZONE)));
+    yesterdayPartition = new Path(partitionPrefix + timeFormatter.format(curTime.minusDays(1)));
+    log.info("Yesterday partition is {}", yesterdayPartition);
+
+    enableEmptyPartition = Boolean.valueOf(properties.getProperty(ENABLE_EMPTY_PARTITION));
+
+    props = properties;
+    this.fs = fs;
+  }
+
+  /**
+   * Derive partition glob pattern from time format. For example:
+   * <p> given {@code timePattern=yyyy/MM/dd}, return '*'/'*'/'*' (adding single quote to avoid breaking comments)
+   * <p> given {@code timePattern=yyyy-MM-dd}, return *
+   */
+  private String derivePartitionPattern(String timeFormat) {
+    String[] parts = timeFormat.split("/");
+    StringBuilder pattern = new StringBuilder("*");
+    for (int i = 1; i < parts.length; i++) {
+      pattern.append("/*");
+    }
+    return pattern.toString();
+  }
+
+  @Override
+  public List<FileSystemDataset> findDatasets()
+      throws IOException {
+    try {
+      return doFindDatasets();
+    } finally {
+      // Recover ConfigurableGlobDatasetFinder.DATASET_FINDER_PATTERN_KEY config
+      this.props.setProperty(ConfigurableGlobDatasetFinder.DATASET_FINDER_PATTERN_KEY, datasetPattern);
+    }
+  }
+
+  private List<FileSystemDataset> doFindDatasets() throws IOException {
+    if (!enableEmptyPartition) {
+      log.info("Will not create empty file system dataset");
+      return findDatasets(datasetPartitionPattern);
+    }
+
+    // Find datasets
+    List<FileSystemDataset> datasets = findDatasets(datasetPattern);
+    // Identify yesterday dataset partitions
+    Set<Path> yesterdayDatasetPartitions = new HashSet<>();
+    datasets.forEach(dataset -> yesterdayDatasetPartitions.add(createYesterdayDatasetPartition(dataset)));
+
+    // Find all dataset time partitions
+    List<FileSystemDataset> datasetPartitions = findDatasets(datasetPartitionPattern);
```

Comment:
Could this code block be optimized to be a single expression like : 
`datasetPartitions.addAll(yesterdaypartitions.stream().allMatch(x->!datasetPartitions.contains(x)).map(x -> createEmptyFileDataset))` ?